src/librbd/operation/DisableFeaturesRequest.h

```diff
@@ -0,0 +1,161 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#ifndef CEPH_LIBRBD_OPERATION_DISABLE_FEATURES_REQUEST_H
+#define CEPH_LIBRBD_OPERATION_DISABLE_FEATURES_REQUEST_H
+
+#include "librbd/operation/Request.h"
+#include "cls/rbd/cls_rbd_client.h"
+
+class Context;
+
+namespace librbd {
+
+class ImageCtx;
+
+namespace operation {
+
+template <typename ImageCtxT = ImageCtx>
+class DisableFeaturesRequest : public Request<ImageCtxT> {
+public:
+  static DisableFeaturesRequest *create(ImageCtxT &image_ctx, Context *on_finish,
+                                       uint64_t features) {
+    return new DisableFeaturesRequest(image_ctx, on_finish, features);
+  }
+
+  DisableFeaturesRequest(ImageCtxT &image_ctx, Context *on_finish,
+			uint64_t features);
+
+protected:
+  virtual void send_op();
+  virtual bool should_complete(int r);
+  virtual bool can_affect_io() const override {
+    return true;
+  }
+  virtual journal::Event create_event(uint64_t op_tid) const {
+    return journal::UpdateFeaturesEvent(op_tid, m_features, false);
+  }
+
+private:
+  /**
+   * DisableFeatures goes through the following state machine:
+   *
+   * @verbatim
+   *
+   * <start>
+   *    |
+   *    v
+   * STATE_BLOCK_WRITES
+   *    |
+   *    v
+   * STATE_APPEND_OP_EVENT (skip if journaling
```

Comment:
I would move this as far down as possible since any error code will result in a mirroring split-brain (e.g. attempting to disable a feature that wasn't enabled).